### ISSUE TITLE
Optional Prop Passing to child components

### ISSUE THREAD

**Issue by davidhewitt** [vuejs/vue#6166]
### What problem does this feature solve?

TLDR; I would like to have a way to "bind" a value to a child component such that the value is passed to the child component as a prop. If the tag is just a native DOM node or the child component doesn't take a prop with that name, nothing happens: value isn't written to the DOM.

The place where this would really help me is as follows:

I have a list template like so:

    <template>
        <component v-for="item in my_items" callback="my_callback_fuction" :is="my_component"/>
    </template>

where my_component can be a number of different components, where some of them have a `callback` prop, and some don't.

Then, in when `my_component` is one that doesn't have a prop named `callback`, we might get emitted DOM like this:

    <li callback="function (x) { ... }"></li>
    <li callback="function (x) { ... }"></li>
    <li callback="function (x) { ... }"></li>
    <li callback="function (x) { ... }"></li>
    <li callback="function (x) { ... }"></li>

The browser serialising `my_callback_function` over and over again is kinda pointless, really bad for performance, and kind of unsatisfying.

I know I can "solve" this by ensuring that all components I might use here have a prop named callback, whether or not they use it. However, the component having to know what props it might need to take feels like a code smell.

### What does the proposed API look like?
I have a couple of ideas about how this could be solved:

1: This problem I think is unique to `<component>`, so perhaps adding a propsData prop to `<component>` would solve the problem.

    
    <component props-data="{callback: my_callback}">`

2: For a more general solution, a modifier could be added to `v-bind`, which passes the value to the child component if it has a property of the given name. If the child component doesn't have such a prop, or is just a native DOM node, the modifier results in a no-op.

Name suggestions:  `v-bind.vue-prop`? `v-bind.prop-data`? `v-bind.child-prop`?

    <component :callback.vue-prop="my_callback">

I am more concerned by a way to have this "optional props passing" than either of these two solutions I offer above. I'd be interested to hear what you think of these and the feature in general. Thanks for considering this.

**Comment by yyx990803** [comment]
If it's just for the purpose of preventing the binding from showing up in the DOM, you can just use `.prop` or the new `inheritAttrs: false` option in 2.4.

**Comment by davidhewitt** [comment]
I'm not sure that `inheritAttrs` is quite what I'm looking for: it's along the right lines but I want to disable it for just one binding rather than for the whole lot.

I'm really confused about the behaviour of `.prop` - I was initially using it like you suggest but in 2.3.0 this changed, and when I reported it in #5925 it seemed to be decided that `.prop` shouldn't be expected to work with child components.

**Comment by yyx990803** [comment]
Apologies, `.prop` modifier would indeed make it not work with components.

`inheritAttrs: false` disables the DOM attribute fallthrough, which seems to solve the exact issue unless you do want to rely on that behavior in some cases? In my experience that's quite rare - most of the time you'd use `class` or `style` which are not affected by `inheritAttrs`. If you do have DOM attributes that you want to pass down, I'd suggest make them props and bind explicitly in the child components.

**Comment by davidhewitt** [comment]
Ah cool, good to know I'm not going insane re `.prop`! I did just spot a weird inconsistency just now though:

    
    <my-input :value.prop="some_value"/>

    
    <component is="my-input" :value.prop="some_value"/>

Here's a fiddle to demonstrate that: https://jsfiddle.net/j1b99p68/. Should this be reported as an issue?

`inheritAttrs: false` - didn't see in the docs that it doesn't affect `class` or `style`? If so, that'll solve what I need to do here, thanks very much @yyx990803!

(bikeshed: It would be nice if somehow this behaviour could be controlled by the parent / user rather than the child needing to know it might have to ignore some stuff, but that's probably a matter of taste!)
